packages/app-contracts/src/ABI.tsx

```diff
@@ -5,65 +5,177 @@
 import { I18nProps } from '@polkadot/ui-app/types';
 
 import React from 'react';
-import { InputFile } from '@polkadot/ui-app';
+import styled from 'styled-components';
+import { InputFile, Labelled, Messages } from '@polkadot/ui-app';
 import { ContractAbi } from '@polkadot/types';
 import { u8aToString } from '@polkadot/util';
 
 import translate from './translate';
 
 type Props = I18nProps & {
-  help: React.ReactNode,
+  contractAbi?: ContractAbi,
+  help?: React.ReactNode,
   isError?: boolean,
-  label: React.ReactNode,
-  onChange: (json: string | null, contractAbi: ContractAbi | null) => void
+  isDisabled?: boolean,
+  isRequired?: boolean,
+  label?: React.ReactNode,
+  onChange: (json: string | null, contractAbi: ContractAbi | null) => void,
+  onRemove?: () => void,
+  onRemoved?: () => void,
+  onSelect?: () => void
 };
 
 type State = {
-  abi?: Uint8Array | null,
+  contractAbi?: ContractAbi,
   isAbiValid: boolean,
-  name?: string,
-  placeholder?: React.ReactNode | null
+  isEmpty: boolean,
+  isError: boolean
 };
 
+const Normalize = styled.div`
+  min-height: 4rem;
+`;
+
 class ABI extends React.PureComponent<Props, State> {
   state: State = {
-    isAbiValid: true
+    isAbiValid: false,
+    isEmpty: true,
+    isError: false
   };
 
+  constructor (props: Props) {
+    super(props);
+    if (props.contractAbi) {
+      this.state = {
+        ...this.state,
+        contractAbi: props.contractAbi,
+        isAbiValid: true,
+        isEmpty: false,
+        isError: false
+      };
+    } else {
+      this.state = {
+        ...this.state,
+        isError: props.isError || props.isRequired || false
+      };
+    }
+  }
+
+  componentWillReceiveProps ({ contractAbi, isError, isRequired }: Props) {
+    if (contractAbi) {
+      this.setState({
+        contractAbi,
+        isAbiValid: true,
+        isError: false
+      });
+    } else if (this.props.contractAbi) {
+      this.setState({
```

Comment:
Yes, should be optional. However, it was added there on purpose, while not needed -

- you have the wasm & abi at that point (compiled)
- once set, you don't need to supply it for each instance

So, this is not about what putCode needs, but rather what makes sense to the user. (Having it required, it not a bad thing)

Reply:
What if you only have the code hash but not the abi, for example if someone just sent you the hash of an existing contract? At the moment you wouldn't be able to add that to the UI.

Reply:
Let’s walk back - what would you like to do with the code hash in this case? You can’t deploy anything from it (since you don’t have the ABI with the deploy call info).

Reply:
I guess the same as when you delete the ABI, which you can do in both the `Code` and the `Contract` section. (I think... I just compiled the latest substrate and so Apps is broken because of the latest metadata changes).
Basically bookmarking it until you get the correct ABI. I'd guess the idea behind being able to delete an ABI had a similar reason: To replace if you accidentally uploaded a wrong ABI...

Reply:
Yes, delete needs a bit of TLC (not required here though, it is already 100x better than what it was), since we really want to always have one available. But it is basically is "editing functionality".